File: tagbag/tag_to_jsonl_converter.py
from .tag_base import TagBase
from .tag_helper import TagHelper


import json
from pathlib import Path
from typing import List, Sequence


class TagToJsonlConverter(TagBase):
    """
    Convert tag files to a JSONL file with fields: image and prompt.
    """

    @staticmethod
    def execute(args: Sequence[str]) -> None:
        """
        args[0]: directory path
        Writes train.jsonl to that directory.
        """
        dir_path = Path(args[0])
        lines: List[str] = []
        for name in TagBase.get_tag_files(dir_path):
            line = TagToJsonlConverter._process_file(Path(name))
            if line:
                lines.append(line)
        (dir_path / "train.jsonl").write_text("\n".join(lines), encoding="utf-8")

    @staticmethod
    def _process_file(file_path: Path) -> str:
        content_lines = file_path.read_text(encoding="utf-8").splitlines()
        if not content_lines:
            return ""
        image_file = TagToJsonlConverter._get_image_file_name(file_path)
        if not image_file:
            return ""
        prompt = " ".join(content_lines)
        obj = {"image": image_file, "prompt": prompt}
        return json.dumps(obj, ensure_ascii=False)

    @staticmethod
    def _get_image_file_name(text_file_path: Path) -> str:
        """
        Find a file in the same directory with the same stem but a different extension.
        Returns empty string if not found.
        """
        directory = text_file_path.parent
        stem = text_file_path.stem
        for f in directory.iterdir():
            if f.stem == stem and f.resolve() != text_file_path.resolve():
                return str(f)
        return ""

File: tagbag/TagToJsonlConverter.cs
namespace NhlStenden.TagBag;

/// <summary>
/// Represents a class that converts tag files to JSONL format.
/// </summary>
public class TagToJsonlConverter : TagBase
{
    /// <summary>
    /// Executes the conversion process on the specified files.
    /// </summary>
    /// <param name="args">A list of arguments where the first argument is the directory path.</param>
    public static void Execute(IList<string> args)
    {
        var fileNames = GetTagFiles(args[0]);
        var lines = new List<String>();
        foreach (var fileName in fileNames)
        {
            var line = ProcessFile(fileName);
            if (string.IsNullOrEmpty(line))
                continue;
            lines.Add(line);
        }
        File.WriteAllLines(Path.Join(args[0], "train.jsonl"), lines);
    }

    /// <summary>
    /// Processes the specified file to convert its content to JSONL format.
    /// </summary>
    /// <param name="fileName">The name of the file to process.</param>
    /// <returns>A JSONL formatted string representing the file content.</returns>
    static string ProcessFile(string fileName)
    {
        var lines = File.ReadAllLines(fileName).ToList();
        if (lines.Count == 0) return "";
        var imageFileName = GetImageFileName(fileName);
        if (string.IsNullOrEmpty(imageFileName)) return "";
        var line = string.Join(" ", lines);
        return "{\"image\": \"" + imageFileName + "\", \"prompt\": \"" + line + "\"}";
    }

    /// <summary>
    /// Retrieves the corresponding image file name for the given text file name.
    /// </summary>
    /// <param name="textFileName">The name of the text file.</param>
    /// <returns>The name of the corresponding image file, if found; otherwise, an empty string.</returns>
    static string GetImageFileName(string textFileName)
    {
        var baseTextFileName = Path.GetFileNameWithoutExtension(textFileName);
        var directory = Path.GetDirectoryName(textFileName);
        if (directory == null) return "";
        var fileNames = Directory.GetFiles(directory, "*.*", SearchOption.TopDirectoryOnly);

        foreach (var fileName in fileNames)
        {
            var baseFileName = Path.GetFileNameWithoutExtension(fileName);
            if (string.Equals(baseFileName, baseTextFileName, StringComparison.Ordinal) && !string.Equals(fileName, textFileName, StringComparison.Ordinal))
                return fileName;
        }

        return "";
    }
}